pkg/arvo/app/notify.hoon

```diff
@@ -0,0 +1,449 @@
+      ::
+          %client-join
+        =/  entry=(unit provider-entry)  (~(get by provider-state) service.act)
+        ?~  entry
+          ~|("no such service: {<service.act>}" !!)
+        ?.  (is-whitelisted:do src.bowl u.entry)
+          ~|("permission denied" !!)
+        =.  clients.u.entry  (~(put by clients.u.entry) src.bowl ~)
+        :_  state(provider-state (~(put by provider-state) service.act u.entry))
+        :~  %:  register-binding:do
+                service.act
+                u.entry
+                binding-endpoint.u.entry
+                src.bowl
+                address.act
+            ==
+            (watch-client src.bowl service.act)
+        ==
+      ::
+          %client-leave
+        =/  entry=(unit provider-entry)  (~(get by provider-state) service.act)
+        ?~  entry
+          ~|("no such service: {<service.act>}" !!)
+        ?.  (is-client:do src.bowl u.entry)
+          ~|("permission denied" !!)
+        =/  client-info=(unit @t)  (~(got by clients.u.entry) src.bowl)
+        =.  clients.u.entry        (~(del by clients.u.entry) src.bowl)
+        :_  state(provider-state (~(put by provider-state) service.act u.entry))
+        ?~  client-info
+          [(leave-client src.bowl service.act)]~
+        :~  %:  remove-binding:do
+                service.act
+                u.entry
+                src.bowl
+                binding-endpoint.u.entry
+                u.client-info
+            ==
+            (leave-client src.bowl service.act)
+        ==
+      ==
+    ::
+    ++  handle-client-action
+      |=  act=client-action
+      ^-  (quip card _state)
+      ?>  (team:title our.bowl src.bowl)
+      ?-  -.act
+          %connect-provider
+        =/  =wire  /connect/(scot %p who.act)
+        =.  providers.client-state
+          (~(put ju providers.client-state) who.act service.act)
+        :_  state
+        [(poke-provider wire who.act %client-join service.act address.act) ~]
+      ::
+          %remove-provider
+        =/  =wire  /remove/(scot %p who.act)
+        =.  providers.client-state
+          (~(del ju providers.client-state) who.act service.act)
+        :_  state
+        [(poke-provider wire who.act %client-leave service.act) ~]
+      ==
+    --
+  ::
+  ++  on-watch
+    |=  =path
+    ^-  (quip card _this)
+    ?+  path  (on-watch:def path)
+        [%notify @ ~]
+      =*  service  i.t.path
+      ?.  (~(has ju providers.client-state) src.bowl service)
+        ~|("permission denied" !!)
+      `this
+    ==
+  ::
+  ++  on-leave
+    |=  =path
+    ^-  (quip card _this)
+    `this
+  ::
+  ++  on-peek   on-peek:def
+  ::
+  ++  on-agent
+    |=  [=wire =sign:agent:gall]
+    ^-  (quip card _this)
+    ?+  wire  (on-agent:def wire sign)
+    ::
+    ::  subscription from client to their own hark-store
+    ::
+        [%hark ~]
+      ?+  -.sign  (on-agent:def wire sign)
+          %fact
+        ?.  ?=(%hark-update p.cage.sign)
```

Comment:
You can factor out `this` into `:_  this` as it isn't modified in any of the cases.